### PR TITLE
[cpu] implement scaled dot product flash attention

### PR DESCRIPTION
**Pull request by Valentine233**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105131
* #107128
* #104863
* #104693
* __->__ #103826
* #104584
* #104583

Feature RFC: https://github.com/pytorch/rfcs/pull/56.

The flash attention CPU kernel is added, for forward path FP32. Blocking is applied on dimensions of query length and kv length and the fusion of gemm + softmax update + gemm is done at once for each block. Parallelization is on the dimensions of batch size, head number and query length. In addition, the causal attention mask is supported. As the attention is masked for the unseen tokens, early termination is applied and we only calculate the blocks in the lower triangular part.

cc @jgong5 @mingfeima @EikanWang @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+#include <ATen/native/transformers/sdp_utils_cpp.h>
+#include <c10/util/irange.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#else
+#include <ATen/ops/empty.h>
+#endif
+
+#if AT_MKL_ENABLED()
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:20]: later on we can remove `mkl.h` from this file, now you can leave it as it is.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+#include <mkl.h>
+#endif
+
+namespace at::native {
+
+namespace {
+
+using fVec = vec::Vectorized<float>;
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:28]: place `using fVec = vec::Vectorized<float>;` in function, do not use global static variable.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+static auto f_vec_size = fVec::size();
+
+inline void _store(
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:31]: we can use `map` from functional to write most of the logic, it would be more clean. as long as the buffer size is L1 hit, the perf doesn't make much difference.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+    int64_t l=fVec::size()) {
+  src.store(dst, l);
+}
+
+inline void _store(
+    at::BFloat16* dst,
+    fVec src,
+    int64_t l=fVec::size()) {
+  auto res = vec::convert_float_bfloat16(src, src);
+  res.store(dst, l);
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:43]: store when size != vec::size() is slow. it has an addional memcpy.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+}
+
+inline void _mkl_gemm(
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:46]: this shall be gone in future.

cpublas::gemm has some additional checks and normalized the 0 stride (some corner cases when tensor is broadcasted). directly go into mkl would have segfault.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+    float* out,
+    float& val) {
+  vec::map<float>(
+    [val](fVec x) { return (x - fVec(val)).exp(); }, out, a, size);
+  val = vec::reduce_all<float>(
+    [](fVec& x, fVec& y) { return x + y; }, out, size);
+}
+
+template <typename scalar_t>
+inline void _normalization_kernel(
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:122]: I will write an overload when the out has different dtype with input in `map`

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+    const int& size,
+    float* out) {
+  float sum_cor = sum_old / sum_new;
+  vec::map<float>(
+    [sum_cor, exp_val](fVec x)
+      { return x * fVec(sum_cor) * fVec(exp_val); },
+      out, a, size);
+}
+
+const std::vector<int64_t> qsplit_range{767, 191, 31};
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:222]: do not use global variables.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+const std::vector<int64_t> qsplit_size{256, 64, 32};
+const int64_t kvsplit_size = 512;
+
+template <typename scalar_t>
+at::ScalarType ConvertTypeToKType() {
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:227]: this did not cover other dtypes such as double.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+    scalar_t* b,
+    float* dst,
+    float* max,
+    float* sum,
+    const float& scale,
+    const int& qsize,
+    const int& kvsize,
+    const int& headsize,
+    const int& idx) {
+  float tmp_max = 0.f, tmp_sum = 0.f, sum_old = 0.f, exp_tmp = 0.f;
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:247]: try to `using accum_t  = at::opmath_type<scalar_t>;` here.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+
+  at::parallel_for(0, batchSize * num_head * qSlice, 0, [&](int64_t begin, int64_t end) {
+    int64_t i = 0;
+    int64_t j = 0;
+    int64_t k = 0;
+    data_index_init(begin, i, batchSize, j, num_head, k, qSlice);
+
+    for (int64_t x = begin; x < end; x++) {
+      int qBlockSize = (k == qSlice - 1) ? qTail : qSplitSize;
+      int ompIdx = at::get_thread_num();
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:328]: move this next to `data_index_init`, this is a slow function call.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+      _init_mha_buffer_kernel(
+          qk_max.data_ptr<float>() + ompIdx * qSplitSize,
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:330]: `data_ptr` is a un-inlined function call, should be placed at inner loop since it is slow.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+  int64_t qSize = query.size(2);
+  int64_t kvSize = value.size(2);
+  int64_t num_head = query.size(1);
+  int64_t headSize = query.size(3);
+  int64_t hiddenSize = num_head * headSize;
+
+  // Query -> Query(Batch x Q_seq_len x hiddenSize)
+  // Key   -> Key(Batch x KV_seq_len x hiddenSize)
+  // Value -> Value(Batch x KV_seq_len x hiddenSize)
+  Tensor q = query.transpose(1, 2).reshape(
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:436]: technically it is possible that this `reshape` may fail when `num_head` and `head_size` are not contiguous dimension.
this is not common case but consider remove the reshape.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 13]
```diff
@@ -0,0 +1,516 @@
...
+      {batchSize, kvSize, hiddenSize});
+
+  int64_t qStride = q.stride(1);
+  int64_t kStride = k.stride(1);
+  int64_t vStride = v.stride(1);
+
+  const auto softmax_scale =
+      sdp::calculate_scale(query, scale).as_float_unchecked();
+
+  if (query.scalar_type() == at::kFloat) {
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:450]: the standard approach here is to use `AT_DISPATCH_FLOATING_TYPES_AND(ScalarType::BFloat16, query.scalar_type(),`

**Review by mingfeima — CHANGES_REQUESTED** [review]
just run some easy to obvious issues, still this patch may need more attention.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+
+  float scaling_factor =
+      sdp::calculate_scale(query, scale).as_float_unchecked();
+
+  // Sizes
+  int64_t batchSize = query.size(0);
+  int64_t qSize = query.size(1);
+  int64_t kvSize = value.size(1);
+  int64_t num_head = query.size(2);
+  int64_t headSize = query.size(3);
```
[the remark this answers is not in the document]
Comment (Valentine233) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:240]: Thanks, updated in PR [add sdpa choice and UT](https://github.com/pytorch/pytorch/pull/105131).

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+  int64_t oStrideM = output.stride(1);
+  int64_t oStrideH = output.stride(2);
+  int64_t lStrideB = logsumexp.stride(0);
+  int64_t lStrideM = logsumexp.stride(1);
+  int64_t lStrideH = logsumexp.stride(2);
+
+  int64_t qSlice = (qSize - 1) / qSplitSize + 1;
+  int64_t num_thread = at::get_num_threads();
+
+  at::Tensor qk = at::empty({num_thread, qSplitSize, kvSplitSize}, at::kFloat);
```
[the remark this answers is not in the document]
Comment (Valentine233) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:262]: Thanks, updated in PR [enable bfloat16 and refactor for flash attention](https://github.com/pytorch/pytorch/pull/104863).

**Review by drisspg — APPROVED** [review]
Going to work my way up the rest of the stack but I think this makes sense, first thought is can we add arbitrary mask support but will continue reading up to see what has changed

**Comment by Valentine233** [comment]
> Going to work my way up the rest of the stack but I think this makes sense, first thought is can we add arbitrary mask support but will continue reading up to see what has changed

Thanks for your reply. For now, the arbitrary mask is not supported in flash attention and this would be added later.

**Review by jgong5 — APPROVED** [review]

**Comment by drisspg** [comment]
@Valentine233  `_scaled_dot_product_efficient_attention` is the aten function that currently accepts an optional attn_bias. Since flash_attention does not support this feature it doesn't have that. If the CPU kernel can support this it might be easier to register this new function under the efficient_attention name.

**Comment by Valentine233** [comment]
> @Valentine233 `_scaled_dot_product_efficient_attention` is the aten function that currently accepts an optional attn_bias. Since flash_attention does not support this feature it doesn't have that. If the CPU kernel can support this it might be easier to register this new function under the efficient_attention name.

Do you mean that for CPU, we do flash attention **WITHOUT** attn_mask in `_scaled_dot_product_flash_attention` and do flash attention **WITH** attn_mask in `_scaled_dot_product_efficient_attention`?

In addition, I am confused why `_scaled_dot_product_flash_attention` does not support this feature.

**Comment by drisspg** [comment]
So currently the only aten function that can calculate SDPA on cpu is _scaled_dot_product_math(). So Flash/MemEff are specifically written to cater to what the cuda implementations support. FlashAttentionV1 and V2 do not support passing in an arbitrary attention bias. Where as of recently: https://github.com/pytorch/pytorch/pull/104310, the kernel does support adding an arbitrary attention bias.

**Comment by drisspg** [comment]
Looking at the CPU implementation it doesn't look there is support for passing an arbitrary attention bias, which is fine just wasn't sure if there was a plan to support this feature

**Comment by Valentine233** [comment]
Sure! In conclusion, if CPU want to add the arbitrary attention bias for flash attention in the future, it should be add in `_scaled_dot_product_efficient_attention`.
